src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -25,7 +26,7 @@
   private static final Logger LOGGER = LoggerFactory.getLogger(Snapshot.class);
   private final String id;
   private final Isolation isolation;
-  private final Map<Key, TransactionResult> readSet;
+  private final Map<Key, Optional<TransactionResult>> readSet;
```

Comment:
Now, empty record is also put into the snapshot to identify reading empty records.